Client/Frontend/Home/BookmarksPanel.swift

```diff
@@ -93,6 +97,54 @@ class BookmarksPanel: SiteTableViewController, HomePanel {
         }
     }
 
+    private func createEmptyStateOverlayView() -> UIView {
+        let overlayView = UIView()
+        overlayView.backgroundColor = UIColor.whiteColor()
+
+        let logoImageView = UIImageView(image: UIImage(named: "emptyBookmarks"))
+        overlayView.addSubview(logoImageView)
+        logoImageView.snp_makeConstraints { make in
+            make.centerX.equalTo(overlayView)
+
+            // Sets proper top constraint for iPhone 6 in portait and for iPad.
+            make.centerY.equalTo(overlayView.snp_centerY).offset(HomePanelUX.EmptyTabContentOffset).priorityMedium()
+
+            // Sets proper top constraint for iPhone 4, 5 in portrait.
+            make.top.greaterThanOrEqualTo(overlayView.snp_top).offset(50).priorityHigh()
+        }
+
+        let welcomeLabel = UILabel()
+        overlayView.addSubview(welcomeLabel)
+        welcomeLabel.text = emptyBookmarksText
+        welcomeLabel.textAlignment = NSTextAlignment.Center
+        welcomeLabel.font = DynamicFontHelper.defaultHelper.DeviceFontLight
+        welcomeLabel.textColor = BookmarksPanelUX.WelcomeScreenItemTextColor
+        welcomeLabel.numberOfLines = 2
+        welcomeLabel.adjustsFontSizeToFitWidth = true
+
+        welcomeLabel.snp_makeConstraints { make in
+            make.centerX.equalTo(overlayView)
+            make.top.equalTo(logoImageView.snp_bottom).offset(BookmarksPanelUX.WelcomeScreenPadding)
+            make.width.equalTo(BookmarksPanelUX.WelcomeScreenItemWidth)
+        }
+        
+        return overlayView
+    }
+
+    private func updateEmptyPanelState() {
+        if source?.current.count == 0 {
+            if self.emptyStateOverlayView.superview == nil {
+                self.tableView.addSubview(self.emptyStateOverlayView)
```

Comment:
I would avoid adding this as a subview to the table view and add it as a subview to the view controller's view but placed above the table view. There isn't necessarily anything wrong with adding it to the table view but I could see issues in the future when table view compatibility since a table view usually controls its subviews/cells using it's delegate.